File: packages/quran-engine-swift/Sources/QuranEngine/Themes.swift
import Foundation

/// Browse the Quran by theme: curated topics, each grouped under a category and a domain, each
/// listing the ayahs that speak to it.
///
/// The ayah lists are curated, not derived, so a topic is a real reading path rather than a keyword
/// hit list. `topicsFor` inverts them. See docs/13-similar-and-themes.md.
public struct Topic: Decodable, Sendable, Equatable {
    public let id: String
    public let name: String
    public let description: String
    public let category: String
    public let domain: String
    /// "2:255" references, in mushaf order.
    public let ayahs: [String]
}

public struct ThemesFile: Decodable, Sendable {
    public let topics: [Topic]
}

public final class Themes {
    private let topics: [Topic]
    private let byId: [String: Topic]
    private var byAyah: [String: [Topic]]?

    public init(_ file: ThemesFile? = nil) {
        self.topics = file?.topics ?? []
        self.byId = Dictionary(topics.map { ($0.id, $0) }, uniquingKeysWith: { first, _ in first })
    }

    /// Every topic, in the order the corpus lists them.
    public func all() -> [Topic] { topics }

    public func topic(_ id: String) -> Topic? { byId[id] }

    /// The distinct domains, in first-seen order.
    public func domains() -> [String] { orderedUnique(topics.map(\.domain)) }

    public func categories(domain: String? = nil) -> [String] {
        let scope = domain.map { d in topics.filter { $0.domain == d } } ?? topics
        return orderedUnique(scope.map(\.category))
    }

    public func inDomain(_ domain: String) -> [Topic] { topics.filter { $0.domain == domain } }

    public func inCategory(_ category: String) -> [Topic] { topics.filter { $0.category == category } }

    /// The topics an ayah appears under.
    public func topicsFor(_ surahId: Int, _ ayahId: Int) -> [Topic] {
        if byAyah == nil {
            var index: [String: [Topic]] = [:]
            for topic in topics {
                for ref in topic.ayahs { index[ref, default: []].append(topic) }
            }
            byAyah = index
        }
        return byAyah?["\(surahId):\(ayahId)"] ?? []
    }

    /// Topics whose name, description, category or domain carries `query`.
    public func search(_ query: String) -> [Topic] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !needle.isEmpty else { return [] }
        return topics.filter { topic in
            [topic.name, topic.description, topic.category, topic.domain]
                .contains { $0.lowercased().contains(needle) }
        }
    }

    private func orderedUnique(_ values: [String]) -> [String] {
        var seen = Set<String>()
        return values.filter { seen.insert($0).inserted }
    }
}

File: packages/quran-engine-js/src/themes.js
// @ts-check
/**
 * Browse the Quran by theme: 323 topics, each grouped under a category and a domain, each listing
 * the ayahs that speak to it.
 *
 * The ayah lists are curated, not derived, so a topic is a real reading path rather than a keyword
 * hit list. `topicsFor` inverts them, which is how an ayah screen can say "this ayah appears under
 * Tawheed and Divine Mercy".
 */

/**
 * @typedef {Object} Topic
 * @property {string} id
 * @property {string} name
 * @property {string} description
 * @property {string} category
 * @property {string} domain
 * @property {string[]} ayahs   "2:255" references, in mushaf order
 */

export class Themes {
  /** @param {{topics: Topic[]}} [data] data/themes.json */
  constructor(data = /** @type any */ ({})) {
    /** @type {Topic[]} */
    this._topics = data?.topics ?? [];
    this._byId = new Map(this._topics.map((t) => [t.id, t]));
    /** @type {Map<string, Topic[]>|null} */
    this._byAyah = null;
  }

  /** Every topic, in the order the corpus lists them. */
  all() {
    return this._topics;
  }

  /** @param {string} id */
  topic(id) {
    return this._byId.get(id) ?? null;
  }

  /** The distinct domains ("Aqeedah (Islamic Creed)", ...), in first-seen order. */
  domains() {
    return [...new Set(this._topics.map((t) => t.domain))];
  }

  /** The distinct categories, optionally within one domain. @param {string} [domain] */
  categories(domain) {
    const scope = domain ? this._topics.filter((t) => t.domain === domain) : this._topics;
    return [...new Set(scope.map((t) => t.category))];
  }

  /** @param {string} domain */
  inDomain(domain) {
    return this._topics.filter((t) => t.domain === domain);
  }

  /** @param {string} category */
  inCategory(category) {
    return this._topics.filter((t) => t.category === category);
  }

  /**
   * The topics an ayah appears under.
   * @param {number} surahId @param {number} ayahId
   * @returns {Topic[]}
   */
  topicsFor(surahId, ayahId) {
    if (!this._byAyah) {
      this._byAyah = new Map();
      for (const topic of this._topics) {
        for (const ref of topic.ayahs) {
          const bucket = this._byAyah.get(ref);
          if (bucket) bucket.push(topic);
          else this._byAyah.set(ref, [topic]);
        }
      }
    }
    return this._byAyah.get(`${surahId}:${ayahId}`) ?? [];
  }

  /**
   * Topics whose name, description, category or domain carries `query`.
   * @param {string} query
   */
  search(query) {
    const needle = query.trim().toLowerCase();
    if (!needle) return [];
    return this._topics.filter((t) =>
      [t.name, t.description, t.category, t.domain].some((f) => f.toLowerCase().includes(needle))
    );
  }
}
